File: app/models/simulation.rb
class Simulation
  extend FFI::Library

  ffi_lib "lib/simulation.so"
  attach_function :c_simulation, [:int], :long

  attr_accessor :input, :ruby_output, :c_output

  def initialize(input)
    @input = input.to_i
  end

  def c_simulation_wrapper
    @c_output = c_simulation(input)
  end

  def ruby_simulation
    output = 0
    1_000.times do |i|
      1_000.times do |j|
        output += i * j * input
      end
    end
    @ruby_output = output
  end
end

File: lib/simulation.c
// Compiled with: gcc -v -O3 -shared -o simulation.so -fPIC simulation.c
#include <math.h>

long c_simulation(int input) {
  long output = 0;
  for (int i = 0; i < 1000; i++) {
    for (int j = 0; j < 1000; j++) {
      output += i * j * input;
    }
  }
  return output;
}
